File: crates/py/examples/ray_casting.py
from common_core_geometry import Point3D, Vector3D, Sphere, Ray, Triangle

def main():
    sphere = Sphere(Point3D(0, 0, 0), 1.0)
    
    width, height = 80, 40
    
    print("Ray-cast sphere rendering:")
    print("=" * width)
    
    for y in range(height):
        row = []
        for x in range(width):
            world_x = (x / width) * 4 - 2
            world_y = (y / height) * 4 - 2
            
            origin = Point3D(world_x, world_y, 5)
            direction = Vector3D(0, 0, -1).normalize()
            ray = Ray(origin, direction)
            
            result = ray.intersect_sphere(sphere)
            if result is not None:
                row.append("█")
            else:
                row.append(" ")
        
        print("".join(row))
    
    print("=" * width)
    
    triangle = Triangle(
        Point3D(-1, -1, 0),
        Point3D(1, -1, 0),
        Point3D(0, 1, 0)
    )
    
    test_ray = Ray(Point3D(0, 0, 5), Vector3D(0, 0, -1).normalize())
    result = test_ray.intersect_triangle(triangle)
    
    if result:
        t, point = result
        print(f"\nRay hit triangle at t={t:.2f}, point={point}")
    else:
        print("\nRay missed triangle")

if __name__ == "__main__":
    main()

File: crates/core/examples/ray_casting.rs
use common_core_geometry::operations::ray_sphere_intersection;
use common_core_geometry::{Point3D, Ray, Sphere, Vector3D};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("=== Ray Casting Example ===\n");

    let ray_origin = Point3D::new(0.0, 0.0, 10.0);
    let ray_direction = Vector3D::new(0.0, 0.0, -1.0);
    let ray = Ray::new(ray_origin, ray_direction)?;

    println!("Ray:");
    println!(
        "  Origin: ({}, {}, {})",
        ray.origin.x, ray.origin.y, ray.origin.z
    );
    println!(
        "  Direction: ({}, {}, {})\n",
        ray.direction.x, ray.direction.y, ray.direction.z
    );

    let spheres = vec![
        (
            Sphere::new(Point3D::new(0.0, 0.0, 0.0), 2.0)?,
            "Center Sphere",
            "red",
        ),
        (
            Sphere::new(Point3D::new(5.0, 0.0, 0.0), 1.5)?,
            "Right Sphere",
            "blue",
        ),
        (
            Sphere::new(Point3D::new(-5.0, 0.0, 0.0), 1.5)?,
            "Left Sphere",
            "green",
        ),
        (
            Sphere::new(Point3D::new(0.0, 5.0, 0.0), 1.0)?,
            "Top Sphere",
            "yellow",
        ),
    ];

    println!("Testing ray intersections:\n");

    for (sphere, name, _color) in &spheres {
        print!("  {}: ", name);
        match ray_sphere_intersection(&ray, sphere) {
            Some((t1, t2)) => {
                let hit1 = ray.point_at(t1);
                let hit2 = ray.point_at(t2);
                println!("HIT!");
                println!(
                    "    First intersection at t={:.2} -> ({:.2}, {:.2}, {:.2})",
                    t1, hit1.x, hit1.y, hit1.z
                );
                println!(
                    "    Second intersection at t={:.2} -> ({:.2}, {:.2}, {:.2})",
                    t2, hit2.x, hit2.y, hit2.z
                );
            }
            None => {
                println!("MISS");
            }
        }
    }

    println!("\n=== Finding closest hit ===\n");

    let mut closest_hit: Option<(f64, &str)> = None;

    for (sphere, name, _color) in &spheres {
        if let Some((t1, _t2)) = ray_sphere_intersection(&ray, sphere) {
            if t1 > 0.0 {
                match closest_hit {
                    None => closest_hit = Some((t1, name)),
                    Some((current_t, _)) => {
                        if t1 < current_t {
                            closest_hit = Some((t1, name));
                        }
                    }
                }
            }
        }
    }

    match closest_hit {
        Some((t, name)) => {
            let hit_point = ray.point_at(t);
            println!("Closest hit: {} at t={:.2}", name, t);
            println!(
                "Hit point: ({:.2}, {:.2}, {:.2})",
                hit_point.x, hit_point.y, hit_point.z
            );
        }
        None => {
            println!("No hits detected!");
        }
    }

    Ok(())
}
